scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1131,6 +1133,47 @@ class MappablePipeJoinEnrichment[T](pipe: TypedPipe[T]) {
 }
 
 /**
+ * This class is to enrich the forceToDisk operation on TypedPipes and allow
+ * us to potentially optimize these operations when they're chained together
+ * redundantly.
+ */
+case class ForceToDiskTypedPipe[T](typedPipe: TypedPipe[T]) extends TypedPipe[T] {
+  import Dsl._
+
+  override def forceToDisk: TypedPipe[T] = this
+
+  override def asPipe[U >: T](fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]): Pipe = {
+    val pipe = typedPipe.toPipe[U](fieldNames)(flowDef, mode, setter)
+    pipe.forceToDisk
+  }
+
+  override def cross[U](tiny: TypedPipe[U]): TypedPipe[(T, U)] =
+    forceBefore(typedPipe).cross(tiny)
+
+  override def flatMap[U](f: T => TraversableOnce[U]): TypedPipe[U] =
+    forceBefore(typedPipe).flatMap(f)
```

Comment:
Yeah for 2, we don't need a new ForceToDiskTypedPipe. We're already ensuring that the rhs has either been forcedToDisk prior to the Cascading hashJoin call or we end up explicitly calling it (when we can verify that).